### PR TITLE
Upgrade Doxia to 2.x stack

### PR DESCRIPTION
**Pull request by slawekjaranowski**

### CODE REVIEW

**Comment by michael-o** [comment]
Will take a look...

**Comment by michael-o** [comment]
There are several problems in `src/main/java/org/codehaus/mojo/taglist/TagListReport.java`:
* The XRef link calcucation should be done like here: https://github.com/apache/maven-pmd-plugin/blob/c27ea6aabdf95797d0f25f1d60875bf485857aa8/src/main/java/org/apache/maven/plugins/pmd/AbstractPmdReport.java#L87-L101 and here https://github.com/apache/maven-pmd-plugin/blob/c27ea6aabdf95797d0f25f1d60875bf485857aa8/src/main/java/org/apache/maven/plugins/pmd/AbstractPmdReport.java#L286-L323
* Locale handling is wrong: https://github.com/mojohaus/taglist-maven-plugin/blob/5bcba292ee22441d2668873acb8bfcdc976998ef/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L148-L151,  https://github.com/mojohaus/taglist-maven-plugin/blob/5bcba292ee22441d2668873acb8bfcdc976998ef/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L528-L534 and other spots. That locale solely refers to the locale of the output. See the abstract parent class.
* You should all methods which are already present in the parent class

`src/main/java/org/codehaus/mojo/taglist/ReportGenerator.java`:
* It should be called `...Renderer` and inherit from `AbstractMavenReportRenderer`
* Maybe move when the move has been done

**Review by michael-o — CHANGES_REQUESTED** [review]
There are several problems in `src/main/java/org/codehaus/mojo/taglist/TagListReport.java`:
* The XRef link calcucation should be done like here: https://github.com/apache/maven-pmd-plugin/blob/c27ea6aabdf95797d0f25f1d60875bf485857aa8/src/main/java/org/apache/maven/plugins/pmd/AbstractPmdReport.java#L87-L101 and here https://github.com/apache/maven-pmd-plugin/blob/c27ea6aabdf95797d0f25f1d60875bf485857aa8/src/main/java/org/apache/maven/plugins/pmd/AbstractPmdReport.java#L286-L323
* Locale handling is wrong: https://github.com/mojohaus/taglist-maven-plugin/blob/5bcba292ee22441d2668873acb8bfcdc976998ef/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L148-L151,  https://github.com/mojohaus/taglist-maven-plugin/blob/5bcba292ee22441d2668873acb8bfcdc976998ef/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L528-L534 and other spots. That locale solely refers to the locale of the output. See the abstract parent class.
* You should all methods which are already present in the parent class

`src/main/java/org/codehaus/mojo/taglist/ReportGenerator.java`:
* It should be called `...Renderer` and inherit from `AbstractMavenReportRenderer`
* Maybe move when the move has been done

**Comment by slawekjaranowski** [comment]
@michael-o  - next part of changes ...

File: src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java [hunk 1]
```diff
@@ -199,12 +162,8 @@ private void doTagDetailedPart(TagReport tagReport) {
...
-        sink.section2();
-        sink.sectionTitle2();
-        sink.anchor(tagReport.getHTMLSafeLinkName());
-        sink.text(tagReport.getTagName());
-        sink.anchor_();
-        sink.sectionTitle2_();
+        startSection(tagReport.getTagName(), tagReport.getHTMLSafeLinkName());
+
         sink.paragraph();
         sink.bold();
```
Comment (slawekjaranowski) [inline src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java:168]: This one use sink direct - there are no methods like start/stop paragraph and also also for bold text.

Alternatively we can use paragraph without bold

Comment (michael-o) [reply src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java]: Keep the bold, if you need bold, no problem.

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java [hunk 2]
```diff
@@ -73,7 +72,7 @@ public class FileAnalyser {
     /**
      * The Locale of the files to analyze.
      */
-    private final Locale locale;
+    private final Locale sourceFileLocale;
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:75]: `sourceFilesLocale`, no?

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java [hunk 3]
```diff
@@ -122,7 +121,7 @@ public FileAnalyser(TagListReport report, List<TagClass> tagClasses) {
         log = report.getLog();
         sourceDirs = report.getSourceDirs();
         encoding = report.getInputEncoding();
-        locale = report.getLocale();
+        sourceFileLocale = report.getSourceFileLocale();
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:124]: From a interface point of view this class should not rely on the `TagListReport` class, but should have independent args/params.

File: src/main/java/org/codehaus/mojo/taglist/TagListReport.java [hunk 4]
```diff
@@ -525,11 +533,7 @@ String getExcludesCommaSeparated() {
      *
      * @return The Locale of the source files.
      */
-    public Locale getLocale() {
-        // The locale string should never be null.
-        if (sourceFileLocale == null) {
-            sourceFileLocale = DEFAULT_LOCALE;
-        }
+    public Locale getSourceFileLocale() {
         return new Locale(sourceFileLocale);
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/TagListReport.java:537]: Note that this one is a bug. The one-arg ctor expects a *language*, not a locale. `de_AT` will fail or result in undefined behavior.  You need something like this: https://github.com/apache/maven-doxia-sitetools/blob/ea6eb76daffa47b4b4fa276fc89b07213c9d7107/doxia-integration-tools/src/main/java/org/apache/maven/doxia/tools/DefaultSiteTool.java#L736

**Review by michael-o — CHANGES_REQUESTED** [review]
Will import the code into my IDE and do another review.

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java (hunk 3 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:124]: you are right ... but for me it is out of scope for this change

File: src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java (hunk 1 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java]: bold is not required .. code looks easier

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java (hunk 2 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:75]: We have a `sourceFileLocale` parameter of Mojo report ... so let to be consistent

File: src/main/java/org/codehaus/mojo/taglist/TagListReport.java (hunk 4 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/mojo/taglist/TagListReport.java]: done

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java (hunk 3 above)
Comment (michael-o) [reply src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:124]: maybe you should add a TODO here...

File: src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java (hunk 2 above)
Comment (michael-o) [reply src/main/java/org/codehaus/mojo/taglist/FileAnalyser.java:75]: Yes, both need to be consistent with a TODO.

File: src/main/java/org/codehaus/mojo/taglist/TagListReport.java [hunk 5]
```diff
@@ -520,17 +528,39 @@ String getExcludesCommaSeparated() {
         }
     }
 
+    void setSourceFileLocale(String sourceFileLocale) {
+        this.sourceFileLocale = sourceFileLocale;
+    }
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/TagListReport.java:533]: You actually don't need the setter unless for testing without DI...

File: src/main/java/org/codehaus/mojo/taglist/TagListReport.java [hunk 6]
```diff
@@ -520,17 +528,39 @@ String getExcludesCommaSeparated() {
...
      */
-    public Locale getLocale() {
-        // The locale string should never be null.
-        if (sourceFileLocale == null) {
-            sourceFileLocale = DEFAULT_LOCALE;
+    public Locale getSourceFileLocale() {
+        String[] items = sourceFileLocale.split("_");
+        if (items.length > 3) {
+            getLog().warn("Invalid java.util.Locale format '" + sourceFileLocale
+                    + "' for 'sourceFileLocale' use ROOT locale");
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/TagListReport.java:544]: use => using

Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/TagListReport.java:544]: Why not the default locale (value) since your default value is `"en"`, not `ROOT`.

**Comment by michael-o** [comment]
Is this on purpose: https://github.com/mojohaus/taglist-maven-plugin/blob/408c9a627f26d672644dd9a11acf1b52e100d651/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L368 
If yes, it should have a comment...

**Comment by michael-o** [comment]
Regarding: https://github.com/mojohaus/taglist-maven-plugin/blob/408c9a627f26d672644dd9a11acf1b52e100d651/src/main/java/org/codehaus/mojo/taglist/TagListReport.java#L638-L641
Though it is not wrong, the canonical way is to use the `I18n` class via DI. See `PMDReport.java`

File: src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java [hunk 7]
```diff
@@ -76,114 +71,82 @@ public class ReportGenerator {
...
      * @param tagReports a collection of tagReports to output.
      */
-    public ReportGenerator(TagListReport report, Collection<TagReport> tagReports) {
-        sortedTagReports = new TreeSet<>(tagReports);
-        this.bundle = report.getBundle();
-        this.sink = report.getSink();
+    public TaglistReportRenderer(TagListReport report, Collection<TagReport> tagReports) {
+        super(report.getSink());
+        this.sortedTagReports = new TreeSet<>(tagReports);
         this.siteOutputDirectory = report.getReportOutputDirectory();
```
Comment (michael-o) [inline src/main/java/org/codehaus/mojo/taglist/TaglistReportRenderer.java:77]: The canonical name is `reportOutputDirectory` see `org.apache.maven.plugin.plugin.report.GoalRenderer`

**Review by michael-o — CHANGES_REQUESTED** [review]
Last round of comments

**Comment by michael-o** [comment]
Also for this plugin:
* https://github.com/apache/maven-reporting-impl/pull/56
* https://github.com/apache/maven-pmd-plugin/pull/172

File: src/main/java/org/codehaus/mojo/taglist/TagListReport.java (hunk 5 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/mojo/taglist/TagListReport.java:533]: it is used in test

**Review by michael-o — APPROVED** [review]
This good enough, let's go.

**Pull request merged**